### PR TITLE
Extra MaybeField Selectors

### PR DESCRIPTION
**Pull request by VitWW**
**Extra MaybeField Selectors**

This proposal add safety and error-less runtime by using `ExtraMaybeFieldSelectors` instead of partial field-selectors.

A bit alternative to #535

[Rendered](https://github.com/VitWW/ghc-proposals/blob/extra-maybe-field-selectors/proposals/0000-extra_maybe_field_selectors.rst)

### CODE REVIEW

File: proposals/0000-extra_maybe_field_selectors.rst [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+Proposed Change Specification
+-----------------------------
+
+Let we have next Sum-Type ``data`` ::
+
+  data Message 
+      = SendChat  { message :: Text,      user :: UserId,  uid :: UserId }
+      | Signup    { eventId :: EventId,   user :: UserId,  uid :: EventId }
+      | SendOrder { signalId :: SignalId, user :: UserId,  uid :: SignalId }
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-extra_maybe_field_selectors.rst]: Oh, it's seems I've over-complicated the proposal by allowing multi-typed fields. 
Ok, I  deleted them!

File: proposals/0000-extra_maybe_field_selectors.rst [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  message  :: Message -> Text     -- or runtime Error
+
+  eventId  :: Message -> EventId  -- or runtime Error
+
+  signalId :: Message -> SignalId -- or runtime Error
+
+  
+  user :: Message -> UserId
+
+  uid  :: Message -> "magical" (UserId | EventId | SignalId)
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-extra_maybe_field_selectors.rst]: Thanks, I removed this.

File: proposals/0000-extra_maybe_field_selectors.rst [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+.. sectnum::
+.. contents::
+
+This proposal add safety and error-less runtime (on wrong use of field selector) by using ``ExtraMaybeFieldSelectors`` instead of partial field-selectors.
+
+
+Motivation
+----------
+
+Partial field-selectors could produce runtime errors and this is extremely weak design. Also duplicated partial field-selectors could have "magical" type of sub-types.
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-extra_maybe_field_selectors.rst]: Thanks, I removed "magical" part of motivation

File: proposals/0000-extra_maybe_field_selectors.rst [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+Import, export
+~~~~~~~~~~~~~~
+
+Sure, we could manually write functions for import and export. We suggest new syntax for field-depended mentioning by adding ``.maybe`` to data-type ::
+
+  module M
+    ( S(x), S(x).maybe
+    , T(..), T(..).maybe
+    ) where
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-extra_maybe_field_selectors.rst:86]: Ok, which details are missing?

**Comment by VitWW** [comment]
@adamgundry  Thank you very much for your opinion and revision!

> Maybe records deserve some slightly ad hoc language support, but it's not obvious how to extend the proposal to updates, let alone something more compositional (like lenses).

I'm not sure about "extending" proposal to lenses. Could you give me an example?

File: proposals/0000-extra_maybe_field_selectors.rst [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+
+Effect and Interactions
+-----------------------
+
+We expect this proposal affects ``OverloadedRecordDot`` extension for maybe-selectors.
+
+We expect this proposal affects ``DisambiguateRecordFields`` extension for reading only maybe-selectors.
+
+We expect this proposal could also affect ``HasField`` class.
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-extra_maybe_field_selectors.rst]: UPDATE for approved [583 HasField Redesign]


```haskell
type HasField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
class HasField x r a | x r -> a where
  -- | Selector function to extract the field @x@ from the record @r@.
  getField :: r -> a
  getField = fromFust . getMaybeField

  getMaybeField :: r -> Maybe a
  getMaybeField = Just. getField

  -- defining just "getField" makes "getMaybeField" same unsafe as "getField"
  {-# MINIMAL getMaybeField | getField #-}



type SetField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
class SetField x r a | x r -> a where
  -- | Change the value stored in the field @x@ of the record @r@.
  modifyField :: (a -> a) -> r -> r
  default modifyField :: (r_rep ~ LiftedRep, a_rep ~ LiftedRep, HasField x r a) => (a -> a) -> r -> r
  modifyField f r = setField @x (f (getField @x r)) r

  modifyMaybeField :: (a -> a) -> r -> Maybe r
  default modifyMaybeField :: (r_rep ~ LiftedRep, a_rep ~ LiftedRep, HasField x r a) => (a -> a) -> r -> Maybe r
  modifyMaybeField f r = (\fn -> setMaybeField @x fn r) <$> (f <$> (getMaybeField @x r))

  -- | Update function to set the field @x@ in the record @r@.
  setField :: a -> r -> r
  -- setField = fromJust . setMaybeField
  default setField :: a_rep ~ LiftedRep => a -> r -> r
  setField v = modifyField @x (\ _ -> v)

  setMaybeField :: a -> r -> Maybe r
  setMaybeField = Just . setField

  -- Haskell DO NOT suport right now several "default" strategies
  -- like "MINIMAL modifyField | setField | modifyMaybeField | setMaybeField"
  -- so now we ignore "MINIMAL setMaybeField | setField | modifyField"
  -- and now we ignore "MINIMAL setMaybeField | setField"
  -- and use instead as was:
  {-# MINIMAL modifyField | setField  #-}
```

**Comment by adamgundry** [comment]
I'd suggest addressing the big picture question of motivation: why is it better to take the approach you propose of having GHC generate extra functions, when the program could use TH to generate them?

The `HasField`/`SetField` classes are a slightly different issue, but open a large number of design questions of their own. With this design, presumably the user just has to know whether to call `getField`or `getMaybeField` and `setField` or `setMaybeField`? There's no way in the type system to determine whether a particular field is partial or total. So I assume `-Wincomplete-record-selectors` would have false positives when using `getMaybeField` (see #516), and similarly for `-Wincomplete-record-updates` and `setMaybeField`. I've thought about possible designs in this area for some time, and I don't currently have a clear solution to propose, but see https://ghc-proposals.readthedocs.io/en/latest/proposals/0583-hasfield-redesign.html#partial-fields for discussion. 

> >  Maybe records deserve some slightly ad hoc language support, but it's not obvious how to extend the proposal to updates, let alone something more compositional (like lenses).
>
> I'm not sure about "extending" proposal to lenses. Could you give me an example?

Currently `r { user = blah }` is partial, and throws an error if `r` is a `SendOrder`. So if we want to get away from partiality involving fields, we need an alternative design of some sort (or we just forbid such updates, e.g. via `-Werror=incomplete-record-updates`). Optics libraries come with affine traversals, which give a nicely compositional and principled story for getting/setting partial fields; but it isn't obvious how to integrate them with Haskell records as they currently exist.

**Comment by VitWW** [comment]
> Currently r { user = blah } is partial, and throws an error if r is a SendOrder. So if we want to get away from partiality involving fields, we need an alternative design of some sort

Thank you! You give an idea how to make my proposal much nicer! 
**I've updated proposal and simplify a lot!**
Now maybe-fileds selectors looks much easier to read! 
`user'maybe :: Message -> Maybe UserId` and we don't care is this field is duplicated or not.

Yes I suggest an alternative design in this proposal! 
We wish to write `r { user'maybe = blah } :: Maybe r`  instead of `r { user = blah } :: r` 
And for several fields update `r { user'maybe = blah, message'maybe = bar } ` we desugars with Maybe-Monad.


> I'd suggest addressing the big picture question of motivation: why is it better to take the approach you propose of having GHC generate extra functions, when the program could use TH to generate them?
> ... There's no way in the type system to determine whether a particular field is partial or total.

Interesting question. I stand on position, that if TH could do this, then we have already a huge library.
Second, people try to avoid to use Template Haskell directly.
Third, this is not about just generating maybe-functions, but also do a lot more, the whole mini ecosystem.
Fourth, this proposal cares if a field-selector is (a) from Record (one Constructor) or Sum-Type (many Constructors) Type ~and (b) is a field-selector is duplicated or not~. This proposal ignore Records. And for this Proposal we don't care if a field-selector is total or partial.

> The HasField/SetField classes are a slightly different issue, but open a large number of design questions of their own. With this design, presumably the user just has to know whether to call getField or getMaybeField and setField or setMaybeField? 

Ok, we could have independent form `HasField`/`SetField` classes: create additional  clases `HasMaybeField`/`SetMaybeField` 
```haskell
class HasMaybeField x r a | x r -> a where
  getMaybeField :: r -> Maybe a

class SetMaybeField x r a | x r -> a where
  modifyMaybeField :: (a -> a) -> r -> Maybe r
  default modifyMaybeField :: (r_rep ~ LiftedRep, a_rep ~ LiftedRep, HasMaybeField x r a) => (a -> a) -> r -> Maybe r
  modifyMaybeField f r = (\fn -> setMaybeField @x fn r) <$> (f <$> (getMaybeField @x r))


  setMaybeField :: a -> r -> r
  default setMaybeField :: a_rep ~ LiftedRep => a -> r -> Maybe r
  setMaybeField v = modifyMaybeField @x (\ _ -> v)

  {-# MINIMAL modifyMaybeField | setMaybeField #-}
```

**Comment by jeukshi** [comment]
> I stand on position, that if TH could do this, then we have already a huge library.

It's not only that TH can do it. We can write it ourselves, now.

```haskell
instance HasField "message'maybe" Message  (Maybe Text) where
  getField :: Message -> Maybe Text
  getField (SendChat msg _) = Just msg
  getField _ = Nothing

f :: Message -> Maybe Text
f msg = msg.message'maybe
```
Yes, it's work we have to do, but that leads me to another point.

> Fourth, this proposal cares if a field-selector is (a) from Record (one Constructor) or Sum-Type (many Constructors) Type

IMO, the programmer should care for which field `<field>'maybe` function even makes sense. If it cannot be specified on per-field basis, then we will end up with a bunch of instances that are not that useful (single field on otherwise big sum type) or with worse ergonomics than pattern matching on constructors (we want to access multiple fields and end up with a bunch of `Maybe`s).

This is something that is missing, for me, in the motivation section. Yes, the proposal eliminates runtime errors, but it does not address why having a `<field>'maybe` selector is better than not having a selector at all.

**Comment by VitWW** [comment]
Good virtual method, I liked it!

> This is something that is missing, for me, in the motivation section. Yes, the proposal eliminates runtime errors, but it does not address why having a `<field>'maybe` selector is better than not having a selector at all.

Short answer is next: a field-selectors hold **meaning** of their values.
Long answer by contradiction is next: if filed-selectors are not handy why Devs did add them into the language? And if filed-selectors are not handy why people do use them defining their own types? 

> IMO, the programmer should care for which field `<field>'maybe` function even makes sense.

I think, if programmer create a field-selector for Sum Type, then this field worth of maybe-field selector

**Comment by jeukshi** [comment]
> Long answer by contradiction is next: if filed-selectors are not handy why Devs did add them into the language? And if filed-selectors are not handy why people do use them defining their own types?

Assuming we are not talking about field-selectors in general, but the sum-type ones. 

I don't know why devs added them back then, but I'd ask a different question: If partial selectors were not in Haskell today, do you think a proposal for adding them would be accepted? I have my doubts about that.

I don't consider partial field-selectors handy; I consider them to be a big foot-gun. I bet some who use them consider them handy, but that is because they haven't realized yet (or simply accepted), that they have been deceived. If I say that a smoke detector is handy, I mean a working one.

Now, if people use them (deceived or not), I don't think it follows, that they would be equally happy with the proposed maybe-selectors. That's what I find lacking in the motivation section. In other words, let's say we live in a world where #535 `NoPartialFieldSelectors` exists. This proposal could simply imply `NoPartialFieldSelectors` and generate maybe-selectors, therefore, this proposal is not about safety at all!

**Comment by VitWW** [comment]
> I don't consider partial field-selectors handy; I consider them to be a big foot-gun.

But this is just a wrong assumption: excluding field-selectors could produce even more errors! Just not runtime-errors, but **logical** errors

Nothing prevents me for creating next functions, but are they safer then with partial selectors? 
```haskell
data Color = RGB Int Int Int | HSL Int Int Int

maybeGreen :: Color -> Maybe Int
maybeGreen (RGB green _ _) = Just green
maybeGreen _ = Nothing

maybeRed :: Color -> Maybe Int
maybeRed (RGB _ _ red) = Just red
maybeRed _ = Nothing

maybeBlue :: Color -> Maybe Int
maybeBlue (HSL _ _ blue) = Just blue
maybeBlue _ = Nothing


data Color = RGB { red :: Int, green :: Int, blue :: Int } 
           | HSL { hue :: Int, saturation :: Int, lightness :: Int }
```

My proposal `ExtraMaybeFieldSelectors` get rid of runtime-errors. Sure, if maybe-selectors (for Sum-Types) are used instead of partial-selectors.

But are maybe-selectors totally safe? 
1) as GET action maybe-selectors are **totally safe** in use! (but partial-selectors are not)
2) as SET/UPDATE action maybe-selectors not cover all variants same as field-selectors, but maybe-selectors hide runtime-errors into `Nothing` values

Is it any cure for SET/UPDATE-action?
_Not in this proposal_.
Let name it `ConstructorInUpdate`, main idea is :
We wish to write instead of `r { green = blah }` both partial-selector(s) and it's Constructor in same scope for checking if they are valid in use, something like
```haskell
r of RGB { green = blah }
```
Weakness of this possible feature - it does not cover duplicated filed selector, like `r of ??? { user = blah }`

**Comment by VitWW** [comment]
I'm closing this proposal in favor for "SubTypes" future Proposal.

Thanks everyone for feedback!

**Pull request closed without merging**